### PR TITLE
irinterp: fix #47250, handle `gc_preserve_[begin|end]`

### LINKED ISSUE #47250: `MethodError(f=Base.string, args=()` when testing QuantumOpticsBase
PkgEval log: https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/c6440b9_vs_d0156b5/QuantumOpticsBase.primary.log

```
State-operator tensor products, sparse |   10     10  1.0s
 Expr(:gc_preserve_begin, Core.Argument(n=3))
Internal error: encountered unexpected error in runtime:
MethodError(f=Base.string, args=(), world=0x000000000000148f)
jl_method_error_bare at /home/kc/julia/src/gf.c:1942
jl_method_error at /home/kc/julia/src/gf.c:1960
jl_lookup_generic_ at /home/kc/julia/src/gf.c:2613 [inlined]
ijl_apply_generic at /home/kc/julia/src/gf.c:2628
error at ./error.jl:44
reprocess_instruction! at ./compiler/ssair/irinterp.jl:229
#_ir_abstract_constant_propagation#515 at ./compiler/ssair/irinterp.jl:312
_ir_abstract_constant_propagation at ./compiler/ssair/irinterp.jl:279
ir_abstract_constant_propagation at ./compiler/ssair/irinterp.jl:413
abstract_call_method_with_const_args at ./compiler/abstractinterpretation.jl:901
abstract_call_method_with_const_args at ./compiler/abstractinterpretation.jl:886
abstract_call_gf_by_type at ./compiler/abstractinterpretation.jl:161
abstract_call_known at ./compiler/abstractinterpretation.jl:1829
abstract_call at ./compiler/abstractinterpretation.jl:1900
abstract_call at ./compiler/abstractinterpretation.jl:1879
abstract_eval_statement_expr at ./compiler/abstractinterpretation.jl:2036
abstract_eval_statement at ./compiler/abstractinterpretation.jl:2251
....
```

### PR DESCRIPTION
**Pull request by aviatesk**
fix #47250

### CODE REVIEW

**Comment by aviatesk** [comment]
@nanosoldier `runtests(["QuantumOpticsBase"], vs = ":master")`

**Comment by nanosoldier** [comment]
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/47252#issuecomment-1285521925) has completed - no new issues were detected. A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/83c037f_vs_f4c4c33/report.html).

**Comment by aviatesk** [comment]
It looks like this PR passes the QuantumOpticsBase test suite.

File: base/compiler/ssair/irinterp.jl [hunk 1]
```diff
@@ -165,15 +165,15 @@ function reprocess_instruction!(interp::AbstractInterpreter,
...
+        cond = inst.cond
+        condval = maybe_extract_const_bool(argextype(cond, ir))
+        if condval isa Bool
             function update_phi!(from::Int, to::Int)
                 if length(ir.cfg.blocks[to].preds) == 0
                     return
                 end
                 for idx in ir.cfg.blocks[to].stmts
                     stmt = ir.stmts[idx][:inst]
-                    isa(stmt, Nothing) && continue
```
Comment (Keno) [inline base/compiler/ssair/irinterp.jl:176]: `nothing` is allowed between PhiNodes, so this condition should be retained.

**Review by Keno — APPROVED** [review]
LGTM modulo the one inline comment

File: base/compiler/ssair/irinterp.jl (hunk 1 above)
Comment (aviatesk) [reply base/compiler/ssair/irinterp.jl:176]: Ops, I missed it. Thanks for your pointer.

**Comment by giordano** [comment]
Gentle reminder that writing "fix #NNN" in the commit/PR title doesn't link the PR to the issue, and won't automatically close the issue when the PR is merged.  The issue number isn't even clickable in the title, it should go in the commit message/PR description: https://github.com/JuliaLang/julia/pull/47227#issuecomment-1284142801.

Also, since this is a bugfix, is there any regression test that can be added?

**Comment by aviatesk** [comment]
> Gentle reminder that writing "fix #NNN" in the commit/PR title doesn't link the PR to the issue, and won't automatically close the issue when the PR is merged. The issue number isn't even clickable in the title, it should go in the commit message/PR description: [#47227 (comment)](https://github.com/JuliaLang/julia/pull/47227#issuecomment-1284142801).

Sure, I will fix up the commit messages.

> Also, since this is a bugfix, is there any regression test that can be added?

It's possible like:
```julia
julia> function semiconcreteeval(s, x)
           local v
           GC.@preserve s begin
               v = s
           end
           v
       end
semiconcreteeval (generic function with 1 method)

julia> code_typed((String,)) do s
           semiconcreteeval(true, s)
       end
Expr(:gc_preserve_begin, Core.Argument(n=2))
ERROR: 
Stacktrace:
  [1] error()
    @ Core.Compiler ./error.jl:44
  ...
```
but hard to make up a sensible one. Most actual code including direct pointer operations won't be semi-concrete interpreted anyway.

File: base/compiler/ssair/irinterp.jl [hunk 2]
```diff
@@ -216,17 +218,18 @@ function reprocess_instruction!(interp::AbstractInterpreter,
...
                 rt = concrete_eval_invoke(interp, inst, mi′, irsv)
             end
-        elseif inst.head === :throw_undef_if_not
-            # TODO: Terminate interpretation early if known false?
+        elseif head === :throw_undef_if_not || # TODO: Terminate interpretation early if known false?
+               head === :gc_preserve_begin ||
+               head === :gc_preserve_end
             return false
         else
             ccall(:jl_, Cvoid, (Any,), inst)
```
Comment (vtjnash) [inline base/compiler/ssair/irinterp.jl:231]: Should normally use `Core.println` in the compiler. The `jl_()` function is normally intended for things that are likely to explode, and gdb.

**Pull request merged**